language/move-prover/bytecode/src/loop_analysis.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    function_data_builder::FunctionDataBuilder,
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    graph::{Graph, Reducible},
+    stackless_bytecode::{
+        BorrowNode,
+        Bytecode::{self, *},
+        Label, Operation, PropKind,
+    },
+    stackless_control_flow_graph::{BlockContent, BlockId, StacklessControlFlowGraph},
+};
+use move_model::{ast, model::FunctionEnv, ty::BOOL_TYPE};
+use std::collections::{BTreeMap, BTreeSet};
+
+#[derive(Debug, Default)]
+pub struct LoopAnnotation {
+    pub loop_targets: BTreeMap<Label, BTreeSet<usize>>,
+}
+
+pub struct LoopAnalysisProcessor {}
+
+impl LoopAnalysisProcessor {
+    pub fn new() -> Box<Self> {
+        Box::new(LoopAnalysisProcessor {})
+    }
+}
+
+impl FunctionTargetProcessor for LoopAnalysisProcessor {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        func_env: &FunctionEnv<'_>,
+        data: FunctionData,
+    ) -> FunctionData {
+        if func_env.is_native() {
+            return data;
+        }
+        let loop_targets = Self::collect_loop_targets(func_env, &data);
+        let mut data = Self::transform(func_env, data, &loop_targets);
+        let loop_annotation = LoopAnnotation { loop_targets };
+        data.annotations.set::<LoopAnnotation>(loop_annotation);
+        data
+    }
+
+    fn name(&self) -> String {
+        "loop_analysis".to_string()
+    }
+}
+
+impl LoopAnalysisProcessor {
+    fn transform(
+        func_env: &FunctionEnv<'_>,
+        mut data: FunctionData,
+        loop_targets: &BTreeMap<Label, BTreeSet<usize>>,
+    ) -> FunctionData {
+        let code = std::mem::take(&mut data.code);
+        let mut builder = FunctionDataBuilder::new(func_env, data);
+        for bytecode in code {
+            match bytecode {
+                Label(attr_id, label) => {
+                    builder.emit(bytecode);
+                    builder.set_loc_from_attr(attr_id);
+                    if loop_targets.contains_key(&label) {
+                        let attr_id = builder.new_attr();
+                        builder.emit(Bytecode::Prop(
+                            attr_id,
+                            PropKind::Assume,
+                            builder.mk_call(
+                                &BOOL_TYPE,
+                                ast::Operation::Not,
+                                vec![builder.mk_call(
+                                    &BOOL_TYPE,
+                                    ast::Operation::AbortFlag,
+                                    vec![],
+                                )],
+                            ),
+                        ));
+                        for idx in &loop_targets[&label] {
+                            let attr_id = builder.new_attr();
```

Comment:
Same here, use`builder.emits_with`. I some cases you may need to use `builder.emits_with(move |id| E)` in case there are expressions build in the context which the closure needs to have ownership of.